### ISSUE TITLE
Batched SVD_LOWRANK being much slower than loop implementation (both CPU and GPU) 

### ISSUE THREAD

**Issue by hypnopump** [pytorch/pytorch#56891]
## Issue description

I've found the `torch.svd_lowrank` to be up to 2x slower in CPU and GPU for the batched implementation when compared to a loop implementation. 

I guess if then the bacthed implemenation could be turned into a loop one so it is faster. 

**Note**: i have found batched and loop implementations to be on par for low matrix sizes (n<2000 for n x n matrices), but very different for big sizes (both in cpu and gpu)

## Code example

```ptyhon
import torch
p = torch.randn(7000, 3)
d = torch.cdist(p, p, p=2)
# optional
# d = d.to(torch.device("cuda:0"))
# loop implementation - faster
u,s,v = [], [], []
for i in range(5):
    u_, s_, v_ = torch.svd_lowrank(d)
    u.append(u_)
    s.append(s_)
    v.append(v_)
u = torch.stack(u, dim=0)
s = torch.stack(s, dim=0)
v = torch.stack(v, dim=0)
# batched implementation - 2x slower
u,s,v = torch.svd_lowrank(torch.stack([d]*5, dim=0))
```

## System Info
I run the script both in a MacBook Pro (torch cpu) and in colab (cuda) with the same results.

**Comment by rgommers** [comment]
Cc @pearu

**Comment by pearu** [comment]
The cause of this phenomenon is in torch.matmul: when the first argument is transposed, matmul is much slower:
```
>>> import time, torch
>>> a, b = torch.randn(5, 7000, 7000), torch.randn(7000, 6)
>>> start = time.time(); c = torch.matmul(a, b); print(time.time() - start)
0.0357511043548584
>>> a, b = torch.randn(5, 7000, 7000).transpose(1, 2), torch.randn(7000, 6)
>>> start = time.time(); c = torch.matmul(a, b); print(time.time() - start)
0.19888591766357422
```

**Comment by pearu** [comment]
It is likely that an extra copy in
https://github.com/pytorch/pytorch/blob/77721ee318d6785010144aa4569efb98199e7162/aten/src/ATen/native/LinearAlgebra.cpp#L1362
makes torch.matmul slower on non-contiguous inputs.

**Comment by pearu** [comment]
Matmul implementation has Fortran strides support, see
https://github.com/pytorch/pytorch/blob/77721ee318d6785010144aa4569efb98199e7162/aten/src/ATen/native/LinearAlgebra.cpp#L887
This means that batched matmul performance can be improved for the given case by relaxing `contiguous()` operation in
https://github.com/pytorch/pytorch/blob/77721ee318d6785010144aa4569efb98199e7162/aten/src/ATen/native/LinearAlgebra.cpp#L1362
to support Fortran contiguous batches. In fact, the following patch could do the job already (untested):
```
Tensor t1 = (tensor1.transpose(-2, -1).is_contiguous()
             ? tensor1
             : tensor1.contiguous()
            ).view({-1, size1[size1.size() - 1]});
```
Task: find a more efficient way to check `tensor1.transpose(-2, -1).is_contiguous()`.

Thoughts?

**Comment by ngimel** [comment]
In this case, it's not Fortran strides vs C strides, it's whether batch dimension can be folded into matrix first dimension (that is , can `(b,m,k)` matrix be viewed as `(b*m, k)`), and that requires C strides. An easy workaround is to send 3d x 3d tensors to matmul, instead of 3d x 2d, that would cause batched matmul to be used, and batched matmul can handle fortran strides
```
a, b = torch.randn(5, 7000, 7000).transpose(1, 2), torch.randn(1,7000, 6)
torch.matmil(a,b)
```

**Comment by jspark1105** [comment]
@ngimel Regarding the smaller A matrix case you shared, I would say this can be considered as a perf issue of batched gemm. Have we shared this with Nvidia?

**Comment by ngimel** [comment]
I agree that's a perf issue with batched gemm, but no, we didn't share it with nv yet.
